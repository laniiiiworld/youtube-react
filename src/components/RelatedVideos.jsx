import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { useYoutubeApi } from '../context/YoutubeApiContext';
import VideoCard from './VideoCard';

export default function RelatedVideos({ keyword }) {
  const { youtube } = useYoutubeApi();
  const {
    isLoading,
    error,
    data: videos,
  } = useQuery(['related', keyword], () => youtube.relatedVideos(keyword), { staleTime: 1000 * 60 * 5 });

  if (isLoading) return <p>Loading...</p>;
  if (error) return <p>Somthing is wrong!</p>;

  return (
    <>
      {videos && (
        <ul>
          {videos.map((video) => (
            <VideoCard key={video.id} video={video} type='list' />
          ))}
        </ul>
      )}
    </>
  );
}
