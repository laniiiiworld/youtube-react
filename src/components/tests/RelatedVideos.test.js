import { render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { withAllContexts, withRouter } from '../../tests/utils';
import { fakeVideos as videos } from '../../tests/videos';
import RelatedVideos from '../RelatedVideos';

describe('RelatedVideos', () => {
  const fakeYoutube = {
    relatedVideos: jest.fn(),
  };

  afterEach(() => fakeYoutube.relatedVideos.mockReset());

  it('renders correctly', async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => videos);
    const { asFragment } = renderRelatedVideos();

    await waitForElementToBeRemoved(() => screen.getByText('Loading...'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders related videos correctly', async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => videos);
    renderRelatedVideos();

    expect(fakeYoutube.relatedVideos).toHaveBeenCalledWith('channelTitle');
    await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(videos.length));
  });

  it('renders loading', () => {
    fakeYoutube.relatedVideos.mockImplementation(() => videos);
    renderRelatedVideos();

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('renders error', async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => {
      throw new Error('error');
    });
    renderRelatedVideos();

    await waitFor(() => expect(screen.getByText('Somthing is wrong!')).toBeInTheDocument());
  });

  function renderRelatedVideos() {
    return render(
      withAllContexts(
        withRouter(
          <Route path='/' element={<RelatedVideos keyword='channelTitle' />} /> //
        ), //
        fakeYoutube
      ) //
    );
  }
});
