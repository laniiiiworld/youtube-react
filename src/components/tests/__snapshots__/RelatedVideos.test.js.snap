// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideos renders correctly 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="cursor-pointer flex flex-row mb-2"
    >
      <img
        alt="title 1"
        class="rounded-xl w-5/12 mr-2"
        src="http://image1/"
      />
      <div
        class="w-7/12"
      >
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title 1
        </p>
        <p
          class="text-sm opacity-80"
        >
          channel title 1
        </p>
        <span
          class="text-sm opacity-80"
        >
          just now
        </span>
      </div>
    </li>
    <li
      class="cursor-pointer flex flex-row mb-2"
    >
      <img
        alt="title 2"
        class="rounded-xl w-5/12 mr-2"
        src="http://image2/"
      />
      <div
        class="w-7/12"
      >
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title 2
        </p>
        <p
          class="text-sm opacity-80"
        >
          channel title 2
        </p>
        <span
          class="text-sm opacity-80"
        >
          just now
        </span>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
